hw/src/ledger.rs

```diff
@@ -0,0 +1,364 @@
+		let mut chunk: [u8; MAX_CHUNK_SIZE] = [0; MAX_CHUNK_SIZE];
+		&mut chunk[0..derivation_path.len()].copy_from_slice(derivation_path);
+		let mut dest_offset = derivation_path.len();
+		let mut data_pos = 0;
+		let mut result;
+		loop {
+			let p1 = if data_pos == 0 { 0x00 } else { 0x80 };
+			let dest_left = MAX_CHUNK_SIZE - dest_offset;
+			let chunk_data_size = min(dest_left, data.len() - data_pos);
+			&mut chunk [dest_offset..][0..chunk_data_size].copy_from_slice(&data[data_pos..][0..chunk_data_size]);
+			result = Self::send_apdu(&handle, commands::SIGN_ETH_TRANSACTION, p1, 0, &chunk[0..(dest_offset + chunk_data_size)])?;
+			dest_offset = 0;
+			data_pos += chunk_data_size;
+			if data_pos == data.len() {
+				break;
+			}
+		}
+
+		if result.len() != 65 {
+			return Err(Error::Protocol("Signature packet size mismatch"));
+		}
+		let v = result[0];
+		let r = H256::from_slice(&result[1..33]);
+		let s = H256::from_slice(&result[33..65]);
+		Ok(Signature::from_rsv(&r, &s, v))
+	}
+
+	fn open_path(&self, path: &str) -> Result<hidapi::HidDevice, Error> {
+		let mut err = Error::KeyNotFound;
+		/// Try to open device a few times.
+		for _ in 0..10 {
+			match self.usb.open_path(&path) {
+				Ok(handle) => return Ok(handle),
+				Err(e) => err = From::from(e),
+			}
+			::std::thread::sleep(Duration::from_millis(200));
+		}
+		Err(err)
+	}
+
+	fn send_apdu(handle: &hidapi::HidDevice, command: u8, p1: u8, p2: u8, data: &[u8]) -> Result<Vec<u8>, Error> {
+		const HID_PACKET_SIZE: usize = 64 + HID_PREFIX_ZERO;
+		let mut offset = 0;
+		let mut chunk_index = 0;
+			loop {
+				let mut hid_chunk: [u8; HID_PACKET_SIZE] = [0; HID_PACKET_SIZE];
+				let mut chunk_size = if chunk_index == 0 { 12 } else { 5 };
+				let size = min(64 - chunk_size, data.len() - offset);
+				{
+					let mut chunk = &mut hid_chunk[HID_PREFIX_ZERO..];
+					&mut chunk[0..5].copy_from_slice(&[0x01, 0x01, APDU_TAG, (chunk_index >> 8) as u8, (chunk_index & 0xff) as u8 ]);
+				
+					if chunk_index == 0 {
+						let data_len = data.len() + 5;
+						&mut chunk[5..12].copy_from_slice(&[ (data_len >> 8) as u8, (data_len & 0xff) as u8, APDU_CLA, command, p1, p2, data.len() as u8 ]);
+					}
+				
+					&mut chunk[chunk_size..chunk_size + size].copy_from_slice(&data[offset..offset + size]);
+					offset += size;
+					chunk_size += size;
+				}
+				trace!("writing {:?}", &hid_chunk[..]);
+				let n = handle.write(&hid_chunk[0..chunk_size])?;
+				if n < chunk_size {
+					return Err(Error::Protocol("Write data size mismatch"));
+				}
+				if offset == data.len() {
+					break;
+				}
+				chunk_index += 1;
+			}
+
+		// read response
+		chunk_index = 0;
+		let mut message_size = 0;
+		let mut message = Vec::new();
+		loop {
+			let mut chunk: [u8; HID_PACKET_SIZE] = [0; HID_PACKET_SIZE];
+			let chunk_size = handle.read(&mut chunk)?;
+			trace!("read {:?}", &chunk[..]);
+			if chunk_size < 5 || chunk[1] != 0x01 || chunk[1] != 0x01 || chunk[2] != APDU_TAG {
```

Comment:
hidapi sends HID report id in the first byte. I've noticed you are using `libusb` for communication. This is a poor choice because it does not allow communicating with HID devices on macOS.  I'm only using `libusb` for hotplug notifications - a feature that your code does not have at all. Perhaps you should be more careful with such accusations.